File: MyScript/edk/c/examples/document/ShapeRecognition.c
/**
 * The `ProcessShapeRecognition` example code shows you how to perform a shape
 * recognition on a document.
 */


// -- configuration ------------------------------------------------------------

// common configuration parameters
#include "../common/config.h"

// include the MyScript Engine API header(s)
#include <c/MyScriptEngine.h>
#include <c/MyScriptDocument.h>

#include <c/MyScriptShape.h>

// include the certificate
#include "../certificates/MyCertificate.h"


// -- example body -------------------------------------------------------------

#include <stdio.h>
#include <stdlib.h>

// global engine (for the sake of readability)
voEngine engine = NULL;

// common utility functions
#include "../common/common.h"

// path to the Document files
// (note that you can rename any notes document as a zip file to watch inside
// it.)
#define TEMPORARY_DIRECTORY OUTPUT_DIR "/tmp"
#define DOCUMENT_FILE       DATA_DIR "/example1.notes"


struct DocumentAsyncCommandResult
{
  bool success;
  voEngineObject object;
};

// Asynchronous content callback (called at completion)
static void contentCompletionCallback(voEngine engine, voDocumentAsyncCommand cmd, bool success, voDocumentAsyncCommandResult result, void* userParam)
{
  struct DocumentAsyncCommandResult* cmdResult = (struct DocumentAsyncCommandResult*)userParam;

  if (cmdResult != NULL)
  {
    cmdResult->success = success;
    cmdResult->object = (success ? voAcquireObject(engine, result.object) : NULL);
  }
}

// main function
int main()
{
  char temporaryDirectoryBuffer[PATH_MAX];
  char documentFileNameBuffer[PATH_MAX];
  voString temporaryDirectory;
  voString documentFileName;

  struct DocumentAsyncCommandResult aSyncResult = { false, NULL };
  TextUIProgressBar progressBar = { 0 };

  voDocument document = NULL;
  voPage page0 = NULL;
  voContent content0 = NULL;

  const voString shapeFieldId = voString_INITIALIZER("myShapeField");
  voContentField shapeField = NULL;

  const voRectangle shapeRectangle = { 160.0f, 661.0f, 1700.0f, 900.0f };
  voPageSelection shapeSelection = NULL;

  voShapeRecognizer shapeRecognizer = NULL;
  voShapeKnowledge shapeKnowledge = NULL;

  voShapeDocument shapeResult = NULL;
  uint32_t segmentCount;

  // -- initialization ---------------------------------------------------------
  if (!loadMyScriptEngine(MYSCRIPT_ENGINE_LIBRARY_DIR, NULL))
    return (fprintf(stderr, "failed to load the MyScriptEngine library\n"), EXIT_FAILURE);

  resolvePath(documentFileNameBuffer, PATH_MAX, DOCUMENT_FILE);
  documentFileName.bytes = documentFileNameBuffer;
  documentFileName.byteCount = strlen(documentFileName.bytes);

  resolvePath(temporaryDirectoryBuffer, PATH_MAX, TEMPORARY_DIRECTORY);
  temporaryDirectory.bytes = temporaryDirectoryBuffer;
  temporaryDirectory.byteCount = strlen(temporaryDirectory.bytes);

  // -- example code -----------------------------------------------------------
  printf("1) creating a new Engine initialized with your certificate\n");
  engine = voCreateEngine(VO_MSE_VER, &myCertificate, NULL);
  verify(engine != NULL);
  printf(" . Engine created successfully\n");
  printf("\n");

  printf("2) loading an existing Document (use default charset)\n");
  document = voLoadObjectFromFile(engine, NULL, &documentFileName, NULL);
  verify(document != NULL);
  verify(VO_IFACE(engine, IDocument)->setTemporaryDirectory(engine, document, NULL, &temporaryDirectory));
  printf(" . ");
  printPath(stdout, documentFileName.bytes);
  printf(" loaded successfully\n");
  printf("\n");

  printf("3) getting the first page of the Document and its content\n");
  page0 = VO_IFACE(engine, IDocument)->getPageAt(engine, document, 0);
  verify(VO_IFACE(engine, IDocument)->sync(engine, document));
  verify(page0 != NULL);
  printf(" . first page retrieved successfully\n");

  aSyncResult.success = false;
  aSyncResult.object = NULL;
  verify(VO_IFACE(engine, IPage)->getContent(engine, page0, contentCompletionCallback, &aSyncResult));
  verify(VO_IFACE(engine, IPage)->sync(engine, page0));
  verify(aSyncResult.success);
  verify(aSyncResult.object != NULL);
  content0 = (voContent)(aSyncResult.object);
  printf(" . first page's content retrieved successfully\n");
  printf("\n");

  printf("4) adding a new shape field to the content of the first page\n");
  shapeField = VO_IFACE(engine, IContent)->addNewField(engine, content0, VO_CONTENT_FIELD_SHAPE, NULL, &shapeFieldId);
  verify(shapeField != NULL);
  printf(" . shape field added successfully\n");
  printf("\n");

  printf("5) selecting a rectangular part of the first page\n");
  shapeSelection = voCreatePageSelection(engine, page0);
  verify(shapeSelection != NULL);
  verify(VO_IFACE(engine, ISelection)->selectRectangle(engine, shapeSelection, &shapeRectangle, VO_SELECT_SET));
  printf(" . selection done successfully\n");
  printf("\n");

  printf("6) adding the selection to the shape field\n");
  verify(VO_IFACE(engine, IContentField)->addItems(engine, shapeField, shapeSelection));
  printf(" . selection added to the shape field successfully\n");
  printf("\n");

  printf("7) creating and configuring a new ShapeRecognizer\n");
  shapeRecognizer = voCreateObject(engine, VO_ShapeRecognizer);
  verify(shapeRecognizer != NULL);
  printf(" . ShapeRecognizer created successfully\n");
  shapeKnowledge = loadEngineObject(SHAPE_KNOWLEDGE_RES);
  verify(shapeKnowledge != NULL);
  printf(" . ShapeKnowledge loaded successfully\n");
  verify(VO_IFACE(engine, IAttachTarget)->attach(engine, shapeRecognizer, shapeKnowledge));
  printf(" . ShapeKnowledge attached to the recognizer successfully\n");
  printf("\n");

  printf("8) performing a recognition on the selection\n");
  initializeTextUIProgressBar(&progressBar);
  verify(VO_IFACE(engine, IContentFieldProcessor)->process(engine, shapeRecognizer, shapeField, updateTextUIProgressBar, &progressBar));
  finalizeTextUIProgressBar(&progressBar);
  printf(" . recognition performed successfully\n");
  printf("\n");

  printf("9) retrieving the recognition result\n");
  shapeResult = VO_IFACE(engine, IContentField)->getResult(engine, shapeField);
  verify(shapeResult != NULL);
  printf(" . result retrieved successfully\n");
  segmentCount = VO_IFACE(engine, IShapeDocument)->getSegmentCount(engine, shapeResult);
  verify(segmentCount != (uint32_t)-1);
  printf(" . retrieved %d segments\n", segmentCount);
  {
    uint32_t segmentId;

    for (segmentId = 0; segmentId < segmentCount; ++segmentId)
    {
      voShapeSegment shapeSegment = NULL;
      uint32_t candidateCount;

      shapeSegment = VO_IFACE(engine, IShapeDocument)->getSegmentAt(engine, shapeResult, segmentId);
      verify(shapeSegment != NULL);

      candidateCount = VO_IFACE(engine, IShapeSegment)->getCandidateCount(engine, shapeSegment);
      verify(candidateCount != (uint32_t)-1);
      printf("   . ShapeSegment #%d has %d ShapeCandidate%s\n", segmentId, candidateCount, ((candidateCount > 1) ? "s" : ""));

      verify(voReleaseObject(engine, shapeSegment));
    }
  }
  printf("\n");

  printf("10) cleaning-up: releasing all created objects and destroying the Engine\n");
  verify(voReleaseObject(engine, shapeResult));
  verify(voReleaseObject(engine, shapeRecognizer));
  verify(voReleaseObject(engine, shapeKnowledge));
  verify(voReleaseObject(engine, shapeSelection));
  verify(voReleaseObject(engine, shapeField));

  verify(voReleaseObject(engine, content0));
  verify(voReleaseObject(engine, page0));
  verify(voReleaseObject(engine, document));
  verify(voDestroyEngine(engine));
  printf(" . clean-up done successfully\n");
  printf("\n");

  printf("end of example:\n");
  printf(" you just learnt how to perform a shape recognition on a document\n");

  // -- finalization -----------------------------------------------------------
  unloadMyScriptEngine();

  return EXIT_SUCCESS;
}

File: MyScript/edk/java/examples/document/ShapeRecognition.java
package document;

import java.io.File;
import java.io.IOException;

import certificates.MyCertificate;

import com.myscript.document.AsyncCommand;
import com.myscript.document.Content;
import com.myscript.document.ContentField;
import com.myscript.document.ContentFieldType;
import com.myscript.document.Document;
import com.myscript.document.IAsyncCommandHandler;
import com.myscript.document.Page;
import com.myscript.document.PageSelection;
import com.myscript.engine.Engine;
import com.myscript.engine.EngineObject;
import com.myscript.geometry.Rectangle;
import com.myscript.ink.SelectionModifier;
import com.myscript.shape.ShapeDocument;
import com.myscript.shape.ShapeKnowledge;
import com.myscript.shape.ShapeRecognizer;
import com.myscript.shape.ShapeSegment;
import common.CFG;
import common.TextUIProgressBar;

/**
 * The <code>ShapeRecognition</code> example code shows you how to perform a shape
 * recognition on a document.
 */
public class ShapeRecognition
{
  /**
   * The temporary directory used in this example.
   */
  private static final String TEMPORARY_DIRECTORY = CFG.OUTPUT_DIR + "/tmp";
  
  /**
   * The document file opened to load its ink. (Note that you can rename any
   * notes document as a zip file to watch inside it.)
   */
  private static final String DOCUMENT_FILE = CFG.DATA_DIR + "/example1.notes";


  /**
   * Main entry point for the <code>ShapeRecognition</code> example program.
   * 
   * @param args command line arguments, ignored.
   * @throws IOException
   */
  public static final void main(String[] args) throws IOException
  {
    if (DOCUMENT_FILE == null || DOCUMENT_FILE.length() == 0)
    {
      System.err.println("/!\\ please choose a valid filename for document file. You can choose document filename by editing DOCUMENT_FILE static member variable of ShapeRecognition class");
      System.exit(-1);
    }
    
    if (TEMPORARY_DIRECTORY == null || TEMPORARY_DIRECTORY.length() == 0)
    {
      System.err.println("/!\\ please choose a valid path for document temporary directory. You can choose the path by editing TEMPORARY_DIRECTORY static member variable of ShapeRecognition class");
      System.exit(-1);
    }

    System.out.println("1) creating a new Engine initialized with your certificate");
    final Engine engine = Engine.create(MyCertificate.getBytes());
    System.out.println(" . Engine created successfully");
    System.out.println();

    System.out.println("2) loading an existing Document");
    final Document document = (Document)EngineObject.load(engine, DOCUMENT_FILE);
    document.setTemporaryDirectory(TEMPORARY_DIRECTORY);
    System.out.println(" . " + new File(DOCUMENT_FILE).getCanonicalPath() + " loaded successfully");
    System.out.println();

    System.out.println("3) getting the first page of the Document and its content");
    final Page page = document.getPageAt(0);
    System.out.println(" . first page retrieved successfully");
    final Content[] contentHandler = new Content[1];
    page.getContent(new IAsyncCommandHandler()
    {
      public void onCompletion(AsyncCommand cmd, boolean success,
                               EngineObject value)
      {
        if (!success)
          throw new RuntimeException("failed to load page content");
        contentHandler[0] = (Content)value.acquire();
      }
    });
    page.sync(); // wait for getContent() to complete
    final Content content = contentHandler[0];
    System.out.println(" . first page's content retrieved successfully");
    System.out.println("");

    System.out.println("4) adding a new shape field to the content of the first page");
    ContentField shapeField = content.addNewField(ContentFieldType.SHAPE,
                                                 "myShapeField");
    System.out.println(" . shape field added successfully");
    System.out.println("");

    System.out.println("5) selecting a rectangular part of the first page");
    final PageSelection selection = PageSelection.create(page);
    final Rectangle shapeRectangle = new Rectangle(160.0f, 661.0f, 1700.0f, 900.0f);
    selection.selectRectangle(shapeRectangle, SelectionModifier.SET);
    System.out.println(" . selection done successfully");
    System.out.println("");

    System.out.println("6) adding the selection to the shape field");
    shapeField.addItems(selection);
    System.out.println(" . selection added to the shape field successfully");
    System.out.println("");

    System.out.println("7) creating and configuring a new shapeRecognizer");
    final ShapeRecognizer shapeRecognizer = ShapeRecognizer.create(engine);
    System.out.println(" . ShapeRecognizer created successfully");
    final ShapeKnowledge shapeKnowledge = (ShapeKnowledge)EngineObject.load(engine, CFG.getShapeKnowledgeRes());
    System.out.println(" . ShapeKnowledge loaded successfully");
    shapeRecognizer.attach(shapeKnowledge);
    shapeKnowledge.dispose();
    System.out.println(" . ShapeKnowledge attached to the recognizer successfully");
    System.out.println("");

    System.out.println("8) performing a recognition on the selection");
    final TextUIProgressBar progressBar = new TextUIProgressBar(System.out);
    progressBar.update(0, 100); // prints 0% status
    shapeRecognizer.process(shapeField, progressBar);
    progressBar.update(100, 100); // prints 100% status
    System.out.println("");
    System.out.println(" . recognition performed successfully");
    System.out.println("");

    System.out.println("9) retrieving the recognition result");
    final ShapeDocument shapeResult = (ShapeDocument)shapeField.getResult();
    System.out.println(" . result retrieved successfully");
    
    final int segmentCount = shapeResult.getSegmentCount();
    System.out.println(" . retrieved " + segmentCount + " segments");
    {
      for (int segmentId = 0; segmentId < segmentCount; ++segmentId)
      {
        ShapeSegment shapeSegment = shapeResult.getSegmentAt(segmentId);
        int candidateCount = shapeSegment.getCandidateCount();
        System.out.println("   . ShapeSegment #" + segmentId + " has " + candidateCount + " ShapeCandidate" + (candidateCount > 1 ? "s" : ""));
        shapeSegment.dispose();
      }
    }
    System.out.println("");

    System.out.println("10) cleaning-up: releasing all created objects and destroying the Engine");
    shapeResult.dispose();
    shapeRecognizer.dispose();
    shapeKnowledge.dispose();
    selection.dispose();
    shapeField.dispose();

    content.dispose();
    page.dispose();
    document.dispose();
    engine.dispose();
    System.out.println(" . clean-up done successfully");
    System.out.println("");

    System.out.println("end of example:");
    System.out.println(" you just learnt how to perform a shape recognition on a document");
  }

} // ShapeRecognition
